Porter Moogle Implementation

MalRD opened the pull request:
Initial implementation of Porter Moogle functionality. The storage functionality of these moogles is implemented by storing toggled bits in the extra data of the "Storage Slip" items that are purchasable from the moogle. Each bit corresponds to if a particular item is or is not stored on the slip. 

Currently, extra data is not (easily) accessible within the Lua code, so functions were placed within lua_baseentity to perform the storage and retrieval of items. Extra data retrieval in Lua should be generalized at some point, perhaps returning the extra bytes as a table of numbers. 

Additionally, the Porter Moogle cannot store items that have started a Magian trial without completing it. Attempting to store one of these items should result in an error message, but currently does not. Since Magian Trials are not yet implemented, this is a low impact defect that can be resolved with the implementation of Magian Trials.

xdemolish commented on scripts/zones/Aht_Urhgan_Whitegate/npcs/Porter_Moogle.lua:
```diff
@@ -0,0 +1,59 @@
+    MAGIAN_TRIAL_ID     =   963
+};
+
+RETRIEVE_DIALOG_ID  =   13462;
```
should probably move this to the zone's TextIDs.lua, and comment the message it displays (for compatibility with the batch is updates)

xdemolish commented on scripts/globals/porter_moogle_util.lua:
```diff
@@ -0,0 +1,337 @@
+        return;
+    end
+
+    if (player:getGil() < 1000) then
```
i *think* this does pretty much the same as the if block below, and the message can probably be moved in an else there instead

MalRD replied:
The previous block was to validate that the event finishing was a retrieve event and not a store / talk. Per my below comment, I'm going to refactor a bit to clarify the intent.

xdemolish commented on scripts/zones/Aht_Urhgan_Whitegate/npcs/Porter_Moogle.lua:
```diff
@@ -0,0 +1,59 @@
+    if (csid == e.TALK_EVENT_ID and option < 1000) then
+        buyStorageSlip(player, option, ITEM_CANNOT_BE_OBTAINED, ITEM_OBTAINED, NOT_HAVE_ENOUGH_GIL);
+    else
+        player:setLocalVar('slipId', 0);
```
shouldn't the var be cleared anyways  (after buying the storage slip)

MalRD replied:
SlipId is set whenever you trade a slip to the moogle for retrieval, so the only time it needs to be cleared is after you exit out of the retrieval menu. 

I'd actually prefer that the clear be inside porter_moogle_util, but while testing scenarios I moved it into Porter_Moogle and built the scripts without correctly it. I'll make a modification and move both handling of both onEventUpdate and onEventFinish back into the utility class.

teschnei commented:
FYI: DSP uses luaJIT, and thus has luaJIT's bitop (http://bitop.luajit.org/) - a whole lot easier to use those instead of reimplementing what they do

helixhamin commented:
Just wanted to mention, it seems like Mog Garden's Porter Moogle is not present.
http://wiki.ffxiclopedia.org/wiki/Porter_Moogle
Not complaining, just noting.

teschnei commented on scripts/globals/porter_moogle_util.lua:
```diff
@@ -0,0 +1,311 @@
+-- desc: Common functionality for Porter Moogles.
+-----------------------------------
+require("scripts/globals/common");
+local bit = require("bit");
```
you can remove this (it's already a global)

teschnei commented on scripts/globals/porter_moogle_util.lua:
```diff
@@ -0,0 +1,311 @@
+        return false;
+    end
+
+    for i = 1, #slipIds, 1 do
```
I think you can just check slipItems[slipId] not nil

and if you do, you can probably condense slipItems and slipIds into one table

teschnei commented on scripts/globals/porter_moogle_util.lua:
```diff
@@ -0,0 +1,311 @@
+-- desc : Begins the storage or retrieval process based on the items
+--        supplied in the trade.
+----------------------------------------------------------------------
+function doTrade(player, trade, e)
```
since these are global, it's probably best to give them a name related to porter moogles directly (porterDoTrade, etc)

xdemolish replied:
'porterDoTrade' wat

porterMoogleTrade
cause there's enough broken engrish on dsp already

MalRD commented:
Alright, I made everything that isn't currently accessed outside of the utilities a local function, renamed the global functions, and corrected a small defect that didn't align with what actually happens on retail.

xdemolish commented on scripts/globals/porter_moogle_util.lua:
```diff
@@ -0,0 +1,305 @@
+-- desc : Begins the storage or retrieval process based on the items
+--        supplied in the trade.
+----------------------------------------------------------------------
+function porterDoTrade(player, trade, e)
```
porterMoogleTrade (since that's what you went with in your npc scripts)

teschnei commented on scripts/globals/porter_moogle_util.lua:
```diff
@@ -0,0 +1,305 @@
+-- desc : Checks if the supplied item is a Moogle Storage Slip.
+----------------------------------------------------------------------
+local function isSlip(itemId)
+	if (slipItems[itemId] ~= nil) then
```
i think this is literally the only check needed in this function - if its not nil, then it is a slip, if it is nil, it isn't

MalRD commented:
Got rid of the silly loop that was there for reasons unknown.

As for the earlier comment about the two table structure: it was done that way to enable an indexed lookup of the itemId for a particular index (slipIds[option]) instead of looping through the second table while keeping track of the iteration. I'm not married to the idea and will change it if you'd prefer a single table.

xdemolish commented on scripts/globals/porter_moogle_util.lua:
```diff
@@ -0,0 +1,297 @@
+            return;
+        end
+        
+        if (player:delGil(1000)) then
```
combine into one check (so player:delGil instead of getGil) cause i get to be picky as well
as far as i could tell, they do like exactly the same shit

if it doesn't work, i will take full responsibility and blame lautan
